### DISCUSSION TITLE
Given an array of TaskEither, what's the best way to run them all in parallel and collect all Left and Right?

### DISCUSSION THREAD

**Discussion by evelant** [gcanti/fp-ts#1730] [Q&A] [answered]
Let's say I have an array of TaskEither but I want to run all of them regardless of failure. What's the best way to achieve that with fp-ts?

I came up with the following:

```
pipe(
    nonCriticalFns,
     A.sequence(T.ApplicativePar),
     T.map(A.separate)
)
```

This gets me the desired `T.Task<Separated<Errors[], Results[]>>` but I'm not sure if there's a better or more concise way to express this. Thoughts?

**Comment by 0x706b** [comment, marked as answer]
You could use `wilt` from the `Witherable` typeclass (basically an effectful `partitionMap`):

```typescript
import { pipe, identity } from 'fp-ts/function'
import * as T from 'fp-ts/Task'
import * as TE from 'fp-ts/TaskEither'
import * as RA from 'fp-ts/ReadonlyArray'

declare const effects: ReadonlyArray<TE.TaskEither<Error, string>>

pipe(
  effects,
  RA.wilt(T.ApplicativePar)(identity),
)
```

**Reply by evelant** [reply]
Thanks! This is a situation where I wish for more documentation as someone still learning FP concepts. I don't think there's any way I could have deduced the purpose of `A.wilt` based on the docs.

The entirety of the docs for it are "`export declare const wilt: PipeableWilt1<'Array'>`". and following that to the `Witherable` docs only gives a single sentence "`Witherable represents data structures which can be partitioned with effects in some Applicative functor.`". Going to the [section on `Wilt`](https://gcanti.github.io/fp-ts/modules/Witherable.ts.html#wilt-interface) from there it's practically impossible to figure out the purpose of `Wilt` from just the very complex type presented.

From my perspective trying to learn these things the docs aren't much help. They tell me that `Witherable` exists but assume that I already understand its purpose and how to use it. Without an example like you just gave it's more or less hopeless for me to have figured out I could use `wilt`.

It would likely be super helpful if there were a 1-2 sentence "plain english" description of each typeclass including "what is this useful for" with a small practical example for each function like you just provided.
